[Distributed] Implement distributed method accessors

xedin opened the pull request:
Make it possible to reach  and call distributed methods via runtime mechanism 
that would look them up by originating method name at runtime and apply with
the given arguments.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[SIL] Add new flag to `SILFunction` - `IsDistributed`

Determines whether given SILFunction represents a distributed
method or its thunk.

xedin committed:
[Mangling] Add mangling for distributed method accessors

`Distributed Method Accessor` is a global helper function
to get access to particular `distributed method` on an actor
from outside the process.

xedin committed:
[IRGen] Add skeleton implementation of distributed method accessor

ktoso commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2193,6 +2193,11 @@ void IRGenSILFunction::emitSILFunction() {
+  // Generate accessor thunk for the `distributed` method.
+  if (CurSILFn->isDistributed() && !CurSILFn->isThunk()) {
+    IGM.emitDistributedMethodAccessor(CurSILFn);
+  }
```
❤️

xedin committed:
[Distributed] IRGen: Add argument loading from `UnsafeRawPointer` into an explosion

`DistributedAccessor::computeArguments` handles loading of the argument
values from the provided `UnsafeRawPointer` (managed by the caller),
coercing retrieved arguments to appropriate platform-native representation,
and exploding object to form correct argument sequence for distributed
method call.

xedin committed:
[Distributed] IRGen: Form a callee for a distributed method

xedin committed:
[Distributed] IRGen: Emit a call to a distributed method

Note that results are not handled yet.

xedin committed:
[Distributed] IRGen: Forward distributed method results through accessor

Accessor infers result(s) from the underlying distributed method its
going to call, and forwards all of the results produced by such call
up to the caller.

ktoso commented on lib/IRGen/GenDistributed.cpp:
```diff
@@ -130,6 +134,97 @@ void IRGenModule::emitDistributedMethodAccessor(SILFunction *method) {
+  auto params = IGF.collectParameters();
+
+  // UnsafeRawPointer that holds all of the argument values
+  auto *argBuffer = params.claimNext();
```
And that's our HeterogenousBuffer then, holding the args?

ktoso replied:
Ah yeah, it is, got it.

xedin committed:
[IRGen] Linking: Make it possible to form async pointers to distributed method accessors

xedin committed:
[Distributed] IRGen: Turn distributed method accessor into async function

xedin committed:
[Distributed] IRGen: Mark accessor as `throws` and forward error values from distributed method

Distributed method is always `async throws` outside of its isolation
context, so accessor has to be `async throws` as well, and be able
to forward error value from the call to the outside.

xedin committed:
[Distributed] IRGen: Support returning results indirectly from accessor

xedin committed:
[SIL] IR: Make sure that distributed thunk is always emitted

Even if there are no uses of the thunk in the code it's still
could be accessed from remotely via distributed accessor mechanism,
so distributed thunks are always used.

xedin committed:
[Distributed] IRGen: Switch accessor over to distributed thunk

Instead of using the distributed method directly, let's always
call distributed thunk instead which is safe option because
actor could not be local (e.g. issue with routing) but only
distributed thunk would know that and take appropriate action.

DougGregor commented on lib/IRGen/GenDistributed.cpp:
```diff
@@ -53,3 +61,294 @@ llvm::Value *irgen::emitDistributedActorInitializeRemote(
+  // through to the distributed method.
+  params.transferInto(arguments, conv.getNumIndirectSILResults());
+
+  // FIXME: Once we remove async task and executor this should be one not three.
```
Async task and executor are long gone... is this a stale FIXME that was copied?

xedin replied:
Yeah, I copied that from `GenThunk.cpp`. I'll remove it from here and there too.

xedin replied:
Fixed!

aschwaighofer reviewed (commented):
Can you add an IRGen test that checks that we emit the distributed thunk, the underlying function , and the remote accessor? And check that the remote accessor calls the distributed thunk.

xedin commented:
@aschwaighofer Yeah, that's the next step for me here, adding IRGen tests for combinations of arguments/result to make sure that everything is generated correctly. Just wanted to figure out whether I'm on the right track here before I continue...

aschwaighofer commented:
Ok. This looks good to me

xedin committed:
[Distributed] NFC: Remove outdated FIXME about async task/executor parameters

xedin committed:
[Distributed] IRGen: Don't load arguments they turn into `void` in native schema

If parameter type's size is `0` then we don't want to perform a load
at the current position in the buffer and have to move to the next
non-empty element.

xedin committed:
[Distributed/Accessor] IRGen: Don't move argument buffer pointer past the last element

xedin committed:
[Distributed/IRGen] NFC: Add various test-cases for distributed accessors

Test multiple scenarios - value types, reference types, optionals,
arrays, direct and indirect returns.

xedin commented on test/IRGen/distributed_actor_accessors.swift:
```diff
@@ -0,0 +1,280 @@
+  distributed func complex(_: [Int], _: Obj, _: String?, _: LargeStruct) -> LargeStruct {
+    fatalError()
+  }
+}
```
@DougGregor and @aschwaighofer Please take a look and let me know what else would be useful to add for the initial concrete-type only implementation. What ever it is has to be able to conform to Codable though...

xedin committed:
[Mangling] Define mangling for runtime accessible function records

xedin committed:
[Distributed] IRGen: Define a runtime record for accessible functions

A "accessible" function that can be looked up based on a string key,
and then called through a fully-abstracted entry point whose arguments
can be constructed in code.

xedin committed:
[IRGen] Implement accessible function section/record emission

Uses a dedicated section in the binary to emit records about
functions that can be looked up by name at the runtime, and
then called through a fully-abstracted entry point whose
arguments can be constructed in code.

xedin committed:
[Distributed] IRGen: Don't allocate offset pointer if there is nothing to load

If underlying function does not have any parameter, then let's
exit early and avoid stack allocating the argument buffer cursor.

xedin committed:
[Distributed] IRGen: Pluralize the name of accessible functions section

xedin committed:
[Distributed] Runtime: Add skeleton support for accessible function sections

xedin committed:
[Distributed] Runtime: Cache previously requested accessible functions

xedin committed:
[Runtime] Expose accessible function lookup as stdlib SPI

`findAccessibleFunction` has to be accessible from Concurrency
and Distributed modules to be able to lookup distributed accessors.

xedin committed:
[Distributed] Runtime: Prototype `executeDistributedTarget` function

xedin committed:
[Distributed] IRGen: Store async function pointer to distributed accessor

Direct pointer to the accessor cannot be called at runtime,
so here is how everything is stored:

- For `distributed` and `async` functions -> async function pointer;
- For regular functions -> function pointer.

xedin committed:
[Distributed] Runtime: Implement `_executeDistributedTarget`

The implementation is as follows:

- Looks up distributed accessor by the given target name
- Extracts information required to setup async context from
  async function pointer stored in accessor record
- Allocates context and calls accessor

xedin committed:
[Distributed] IRGen: Adjust distributed method accessor to store result into a provided buffer

Instead of trying to return result from distributed thunk directly,
modify accessor to store result into the caller-provided buffer.
Doing so helps us avoid boxing the result into `Any`.

xedin committed:
[Distributed] Remove outdated declaration of `_executeDistributedTarget` from `DistributedActor.swift`

The function should be (and is) declared in `DistributedActorSystem.swift`

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please clean test macOS platform

xedin commented:
@swift-ci please clean test macOS platform

xedin commented:
@swift-ci please clean test macOS platform

aschwaighofer reviewed (approved):
Looks good to me.

xedin committed:
[Distributed] NFC: Update accessor test-cases to use `FakeActorSystem` instead of `Transport`

xedin commented:
@swift-ci please clean test

ktoso reviewed (approved):
I was reviewing this but don't have much to add other than that it looks good and I hope to connect it all up shortly :-)

xedin commented on include/swift/Runtime/AccessibleFunction.h:
```diff
@@ -0,0 +1,34 @@
+
+SWIFT_RUNTIME_STDLIB_SPI const AccessibleFunctionRecord *
+swift_findAccessibleFunction(const char *targetNameStart,
+                             size_t targetNameLength);
```
We might use flags to detect new scheme of naming if we switch to something other than mangled names.

ktoso replied:
My worry is about both "sides" having to use the same scheme so they'd have to be compiled using the same flags, but I guess that's just what we'll live with 🤔

xedin replied:
Yeah, that's the unfortunate reality of the situation.

ktoso replied:
Heh yeah... that's what I kept bringing up as "we change it later" being a problem... so we'll be forever stuck emitting the mangled type names as identifiers, and may be able to–in addition to those–add some shorter names... Unless we know "all sides" of the peers and can decide on a scheme to use with a compilation flag... I guess there's ways out of this corner but it'll be annoying :-/

xedin replied:
Yeah, I think if we just wanted to drop all of the type information then it would be pretty easy to handle at runtime because we could just read mangled name from the record, remove the type information and match against the name provided by the caller. We'd only have to do that transformation once per method and cache it, but for other schemes it would be tricker.

ktoso replied:
Hmmm, I wonder if we could get away with recording using the "full name" and then compare the mangled name if we have it to disambiguate any potential stores with the same name... 🤔 


Not a blocker for merging this, but just something to continue thinking about L(

ktoso commented on test/Distributed/distributed_actor_accessors.swift:
```diff
@@ -0,0 +1,381 @@
+
+/// -> `MyActor.simple1`
+// CHECK:      @"$s27distributed_actor_accessors7MyActorC7simple1yySiFTEHF" = private constant
+// CHECK-SAME: @"symbolic Si___________pIetMHygzo_ 27distributed_actor_accessors7MyActorC s5ErrorP"
```
I just realized... `___` can be used to just ignore bits of text!?

ktoso replied:
Ok not really, that's the actual text. :)

xedin replied:
It's the actual text, sorry, FileCheck cannot do that expect maybe by pattern matching

xedin committed:
[Distributed] Adjust accessor tests to match Windows/Linux section name correctly

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

xedin committed:
[Distributed] IRGen: Load arguments directly into argument explosion

It's incorrect to load from the pointer directly, instead accessor
should use `loadAsCopy` for loadable types.

xedin committed:
[Distributed] IRGen: Make sure that arguments extracted from buffer have correct ownership

Use parameter convention information to determine ownership of the extracted arguments.

xedin committed:
[Distributed] NFC: Remove `align 8` from the accessor thunks test-cases

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please clean test macOS platform

xedin committed:
[Distributed] Runtime: Fix accessor section test-case to support 32/64 bit platforms

xedin committed:
[Distributed] Runtime/NFC: Split accessor thunk tests into 32/64 bit versions

xedin committed:
[Distributed] Temporary disable accessor tests on Windows

I don't have access to Windows machine so there is no way for
me to debug and fix them.

xedin commented:
@swift-ci please clean test

aschwaighofer reviewed (approved):
lgtm
